### DISCUSSION TITLE
Error when calling a method within the "then" clause of another mthod

### DISCUSSION THREAD

**Discussion by JohnLloydJones** [alpinejs/alpine#4089] [7. Bugs (archived)] [answered]
I have a method, loadTree, that get called like this:
              &lt;div class="dirtree" x-init="loadTree()">
					&lt;ul>
						&lt;template x-for="(level,i) in levels">
							&lt;li x-html="renderLevel(level,i)"></li>
						&lt;/template>
					&lt;/ul>
				&lt;/div>
			&lt;/div>

This works fine with no errors. I have another method, search, that gets triggered by a user action and performs a service call to set search parameters and then, in the ".then" block, calls loadTree(). Maddeningly, the UI gets updated as expected, but the console displays a flurry of errors. The exact point of error is (using alpine 3.13.7) here:
  // packages/alpinejs/src/directives/x-ref.js
  function handler3() {
  }
  handler3.inline = (el, { expression }, { cleanup: cleanup2 }) => {
    let root = closestRoot(el);
    if (!root._x_refs)
      root._x_refs = {};
    root._x_refs[expression] = el;
    cleanup2(() => delete root._x_refs[expression]);
  };
  directive("ref", handler3); 

The method closestRoot(el); can return a void value and the code does not handle the case when root == 'undefined'. If I simply  test the value before using it : {{  if (!root) return;  }} my application works correctly with no errors in the console. Of course, I have no idea what else I may have broken as my app is still quite simple. 

This attached file is my application javascript
[tree.js.txt](https://github.com/alpinejs/alpine/files/14553632/tree.js.txt)

**Comment by ekwoka** [comment, marked as answer]
You really just shouldn't be using refs here at all.

I'm very confused how it is managing to not have a root though in this context... That DOES seem like a bug, but I think here it's more that the error exposes that what you're doing is probably not smart.

You shouldn't be doing this direct manipulation of styles. 

```diff
- el.style.display = 'none'
+ x-show="level.show"
```

For example

You should model the states (level is open/shown closed/hidden) in the data itself, and using directives to bind that to the UI.

This should get past these issues while making the code and stuff more clear.

I'd also say that `x-html` is not a good choice almost always, but this is one case in which it or something else like it is kind of required (for nested templates at an arbitrary depth)

**Comment by JohnLloydJones** [comment]
Thanks for the rapid response. I'm still on a steep learning curve and the use of refs was gleaned from the web. I'll refactor my code after your suggestions.

**Reply by ekwoka** [reply]
[You can check this out for reference](https://awesomealpine.com/play?coreplugins=2&script=QWxwaW5lLmRhdGEoJ3RyZWUnLCAoKSA9PiAoewogIHNob3dIaWRkZW46IGZhbHNlLAogIHRyZWU6IHsKICAgIG5hbWU6ICdyb290JywKICAgIGRpcjogdHJ1ZSwKICAgIGZpbGVzOiBbCiAgICAgIHsKICAgICAgICBuYW1lOiAnZG9jdW1lbnRzJywKICAgICAgICBkaXI6IHRydWUsCiAgICAgICAgZXhwYW5kZWQ6IGZhbHNlLAogICAgICAgIGZpbGVzOiBbCiAgICAgICAgICB7CiAgICAgICAgICAgIG5hbWU6ICdqYXZhc2NyaXB0JywKICAgICAgICAgICAgZGlyOiBmYWxzZSwKICAgICAgICAgIH0sCiAgICAgICAgICB7CiAgICAgICAgICAgIG5hbWU6ICd0eXBlc2NyaXB0JywKICAgICAgICAgICAgZGlyOiBmYWxzZSwKICAgICAgICAgIH0sCiAgICAgICAgXSwKICAgICAgfSwKICAgICAgewogICAgICAgIG5hbWU6ICdnaXQnLAogICAgICAgIGRpcjogdHJ1ZSwKICAgICAgICBleHBhbmRlZDogZmFsc2UsCiAgICAgICAgZmlsZXM6IFsKICAgICAgICAgIHsgbmFtZTogJy56c2hyYycsIGRpcjogZmFsc2UsIGhpZGRlbjogdHJ1ZSB9LAogICAgICAgIF0sCiAgICAgIH0sCiAgICBdLAogIH0sCiAgcmVuZGVyKGxldmVsKSB7CiAgICBpZiAobGV2ZWwuZGlyKQogICAgICByZXR1cm4gYAogICAgICA8c3BhbiB4LXRleHQ9Iml0ZW0ubmFtZSI-PC9zcGFuPgogICAgICA8aW5wdXQgdHlwZT0iY2hlY2tib3giIHgtbW9kZWw9Iml0ZW0uZXhwYW5kZWQiPgogICAgICA8dWwgeC1zaG93PSJpdGVtLmV4cGFuZGVkIiB4LWNvbGxhcHNlPgogICAgICA8dGVtcGxhdGUgeC1mb3I9Iml0ZW0gaW4gaXRlbS5maWxlcyI-CiAgICAgICAgPGxpIHgtaHRtbD0icmVuZGVyKGl0ZW0pIiBjbGFzcz0ibWwtNCIgeC1zaG93PSJzaG93SGlkZGVuIHx8ICFpdGVtLmhpZGRlbiIgeC1jb2xsYXBzZT48L2xpPgogICAgICAgIDwvdGVtcGxhdGU-CiAgICAgICAgPHVsPmA7CiAgICByZXR1cm4gYDxzcGFuIHgtdGV4dD0iaXRlbS5uYW1lIj48L3NwYW4-YDsKICB9LAp9KSk7Cg&tw=1&html=PGRpdiB4LWRhdGE9InRyZWUiPgogIDxidXR0b24KICAgIHR5cGU9ImJ1dHRvbiIKICAgIEBjbGljaz0ic2hvd0hpZGRlbiA9ICFzaG93SGlkZGVuIgogID4KICAgIFRvZ2dsZSBIaWRkZW4gPHNwYW4geC10ZXh0PSJzaG93SGlkZGVuIj48L3NwYW4-CiAgPC9idXR0b24-CiAgPHRlbXBsYXRlIHgtZm9yPSJpdGVtIGluIFt0cmVlXSI-CiAgICA8ZGl2IHgtaHRtbD0icmVuZGVyKGl0ZW0pIj48L2Rpdj4KICA8L3RlbXBsYXRlPgo8L2Rpdj4K&panel=script)

This has nested directories, hidding and showing elements, collapsing folders, etc. (might want to build in something that will limit the depth...)

GENERALLY, `refs` should be avoided. The vast majority of the time, directly manipulating the dom, or using refs is a code-smell that the approach being used is the wrong one.

Not all the time, but it's a smart rule to use on first approaching a problem.

**Reply by JohnLloydJones** [reply]
I got rid of the refs and my new code is much cleaner and simpler.  It works without any errors. Many thanks for your help.
